### ISSUE TITLE
Improve handling of sandbox URLs when a username changes

### ISSUE THREAD

**Issue by ragesoss** [WikiEducationFoundation/WikiEduDashboard#4276]
### Current Behavior:
When a username changes and that user has Assignments, the sandbox_url values will remain pointed to the userspace of the old username. This can affect both the user whose name changed, as well as classmates who are assigned to the same group or are peer reviewing that user's articles.

### Desired Behavior:
Upon username change:

* Identify which sandboxes corresponding to that student's Assignments, as well as classmates' peer reviews, have already been created in the old userspace, and perform a page move to bring them to the new userspace of the renamed user.
* Update the sandbox_url for each Assignment in the course that pointed to the old username, replacing it with the new one.

### Related Files & Code:

`UserImporter.from_omniauth` is where the system changes an updated username upon login. This is probably the place to begin.

**Comment by BaileyMcKelway** [comment]
Is anyone working on this? I don't have much experience with Rails but figured this would be a good way to get into it.

**Comment by ragesoss** [comment]
Nope, no one is working on it. And yes, I think this is a great issue to serve as an intro to Rails.

**Comment by BaileyMcKelway** [comment]
Ok great! Going to start on this.

**Comment by RinkiyaKeDad** [comment]
Hey @BaileyMcKelway, are you still working on this? If yes would you like some help? I'm new to Rails too and would love to contribute :D

**Comment by BaileyMcKelway** [comment]
Heyo 👋

I believe I have found a solution, I'm just being slow since it's the holidays.

One thing I have found is that it is a bit tricky to test my solution since that involves changing usernames via MediaWiki; which is not instantaneous.

I could really use some help testing my solution. Next week I'll create a PR and we can test it.

**Comment by ragesoss** [comment]
Cool. It's tricky to do (in particular) browser tests that involve the OAuth login features — although I think it can be done. It's relatively easy to simulate a username change for manual testing, though: just update your user record from the Rails console to change the username. Whatever you change it to will act as the "old" username, and when you next log in to the dashboard, it will go through that UserImporter code to update the username to the real one. This should be basically equivalent to what happens when you actually change your username on the mediawiki side of things. The `global_id` from mediawiki is what stays the same through username changes, and the dashboard uses that to look up the User record if searching by username fails during login.

**Comment by BaileyMcKelway** [comment]
Does this look right?

This is the function in user_importer. I added lines 88-96
![Screen Shot 2020-12-02 at 1 11 47 PM](https://user-images.githubusercontent.com/25776618/100913562-f16f9c00-349f-11eb-9cf7-54c6dfdc63ab.png)


Also from the description: 
> Identify which sandboxes corresponding to that student's Assignments, as well as classmates' peer reviews, have already been created in the old userspace, and perform a page move to bring them to the new userspace of the renamed user.

I couldn't any models handling peer reviews.

**Comment by ragesoss** [comment]
On a quick glance, that looks like you're on the right track.

Peer reviews are also represented by the Assignment model; the `role` attributed determines whether it's an article the user is assigned to write (role 0) or to peer review for a classmate (role 1).

The key thing here is to find all the assignments for the whole class have the old username within the sandbox_url. The corresponding logic for why various assignments in the class (including ones for different users) can point to the same sandbox_url is in `Assignment#set_sandbox_url`. Basically, the first user in the class who gets assigned a particular article determines which sandbox_url will be used by anyone else working on the same article (so that students who are working together on the same article can collaborate in the same sandbox, and classmates providing peer review feedback can find the sandbox where the assigned user has drafted their work for review).

**Comment by BaileyMcKelway** [comment]
![Screen Shot 2020-12-03 at 6 58 33 PM](https://user-images.githubusercontent.com/25776618/101104637-90c48a00-3599-11eb-8f04-12f7803feb86.png)

This seems to be working

**Comment by BaileyMcKelway** [comment]
I was trying to find a way to do it with [Active Record Callbacks](https://api.rubyonrails.org/classes/ActiveRecord/Callbacks.html) in the Model but couldn't find a simple solution.

**Comment by ragesoss** [comment]
Hmm... I would usually want to avoid anything this extensive in a callback, although I see the appeal of it.

What was your approach to a callback here? I would think the appropriate callback would be in the `User` model, where if you save a record with a changed username, it triggers this stuff to happen. The benefit is that it's a more thorough guarantee that this update will happen whenever a username changes, without assuming that `UserImporter.update_user_from_auth` is the only way that a username change can happen (even though, currently, it is). The downside is that the callback code would run every time a User record is saved.

In either case, this routine should definitely be extracted to its own file. I think we'll want to extend the routine at some point to do more things — check whether any of the old sandbox_urls have content in them already, and if so, maybe automatically rename the wiki pages to the new urls — which is another point in favor of not using a callback approach.

**Comment by Abdulwahab03** [comment]
Hello @ragesoss, I'm new here and I would like to take up this issue.
I dropped message to add me to the group Slack channel,  this is my mail wsaheed77@gmail.com

**Comment by ragesoss** [comment]
> Hello [@ragesoss](https://github.com/ragesoss), I'm new here and I would like to take up this issue. I dropped message to add me to the group Slack channel

Invite sent!

**Comment by nexpectArpit** [comment]
@ragesoss is this issue open for contribution?

**Comment by ragesoss** [comment]
@nexpectArpit yes

**Comment by AarohiSharma5** [comment]
Hi @ragesoss, I inspected this issue. I see, the sandbox_url stays the same even when the username is changed. Once, we try to fetch the assignments, we can only access them through the sandbox_url using the old username. I also explored the methods involved in the updation and redirection, and found out "self.update_user_from_auth" method present in the user_importer.rb only focuses on updating the user table. It doesn't update the sandbox URLs in the assignments table. I think of adding more logic to it, by checking if the old username matches the "auth.info.name". If not, we can introduce a new method- something like, if not found, updating the sandbox url in the assignments table. Do I make sense to you? What do you think about it? Shall I move forward with this?

**Comment by nexpectArpit** [comment]
@AarohiSharma5 i am working on this!

**Comment by AarohiSharma5** [comment]
Oh okay- 
I didn't see anyone claiming to work on it, so started working on it..
